### PR TITLE
Feat: Configurable RPC Apis

### PR DESCRIPTION
**Pull request by crypto-facs**
Closes: #174 

## Description
Given that we now support some debug endpoints. It is necessary to allow nodes to enable the desired api namespaces to avoid undesired stress on the nodes from expensive requests.

This PR just enables to specify which namespaces to have open from the namespaces that we support. Left eth namespace open by default.

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: docs/basics/json_rpc.md [hunk 1]
```diff
@@ -11,6 +11,16 @@ Check the JSON-RPC methods and namespaces supported on Ethermint. {synopsis}
 - [Ethereum JSON-RPC](https://eth.wiki/json-rpc/API) {prereq}
 - [Geth JSON-RPC APIs](https://geth.ethereum.org/docs/rpc/server) {prereq}
 
+## JSON-RPC Server
+
+To enable RPC server use the following flag (set to true by default).
+
+```ethermintd start --grpc.enable```
```
Comment (yihuang) [inline docs/basics/json_rpc.md:18]: "evm-rpc" here?

File: ethereum/rpc/apis.go [hunk 2]
```diff
@@ -33,53 +32,80 @@ const (
...
-		{
-			Namespace: DebugNamespace,
-			Version:   apiVersion,
-			Service:   debug.NewInternalAPI(ctx),
+			Service:   filters.NewPublicAPI(ctx.Logger, tmWSClient, evmBackend),
 			Public:    true,
 		},
+	)
+
+	for _, api := range selectedAPIs {
```
Comment (yihuang) [inline ethereum/rpc/apis.go:60]: Do we need to deduplicate here?

File: ethereum/rpc/apis.go [hunk 3]
```diff
@@ -33,53 +32,80 @@ const (
...
-		{
-			Namespace: Web3Namespace,
-			Version:   apiVersion,
-			Service:   web3.NewPublicAPI(),
-			Public:    true,
-		},
-		{
+	var apis []rpc.API
+
+	//Eth apis will be set by default
```
Comment (thomas-nguy) [inline ethereum/rpc/apis.go:42]: Do you mean this cannot be disable?

Comment (yihuang) [reply ethereum/rpc/apis.go:42]: there's `evm-rpc.enabled` already I think.

Comment (thomas-nguy) [reply ethereum/rpc/apis.go:42]: ok good enough

**Review by leejw51crypto — APPROVED** [review]
lgtm

File: CHANGELOG.md [hunk 4]
```diff
@@ -71,6 +71,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (rpc) [tharsis#176](https://github.com/tharsis/ethermint/issues/176) Support fetching pending nonce
 * (rpc) [tharsis#272](https://github.com/tharsis/ethermint/pull/272) do binary search to estimate gas accurately
 * (rpc) [#313](https://github.com/tharsis/ethermint/pull/313) Implement internal debug namespace (Not including logger functions nor traces).
+* (rpc) [#313](https://github.com/tharsis/ethermint/pull/313) Implement internal debug namespace (Not including logger functions nor traces).
```
Comment (fedekunze) [inline CHANGELOG.md:74]: update

File: cmd/ethermintd/config/config.go [hunk 5]
```diff
@@ -57,27 +57,30 @@ func DefaultConfig() *Config {
...
-	}
-}
-
 // EVMRPCConfig defines configuration for the EVM RPC server.
 type EVMRPCConfig struct {
 	// RPCAddress defines the HTTP server to listen on
 	RPCAddress string `mapstructure:"address"`
 	// WsAddress defines the WebSocket server to listen on
 	WsAddress string `mapstructure:"ws-address"`
+	// Api defines the api namespaces that should be enabled
```
Comment (fedekunze) [inline cmd/ethermintd/config/config.go:66]:
```suggestion
	// API defines a list of JSON-RPC namespaces that should be enabled
```

File: cmd/ethermintd/config/config.go [hunk 6]
```diff
@@ -57,27 +57,30 @@ func DefaultConfig() *Config {
...
+	API string `mapstructure:"api"`
```
Comment (fedekunze) [inline cmd/ethermintd/config/config.go:67]: I think this should be a slice
```suggestion
	API []string `mapstructure:"api"`
```

File: cmd/ethermintd/config/config.go [hunk 7]
```diff
@@ -57,27 +57,30 @@ func DefaultConfig() *Config {
...
 	Enable bool `mapstructure:"enable"`
 	// EnableUnsafeCORS defines if CORS should be enabled (unsafe - use it at your own risk)
 	EnableUnsafeCORS bool `mapstructure:"enable-unsafe-cors"`
 }
 
+// DefaultEVMConfig returns an EVM config with the JSON-RPC API enabled by default
+func DefaultEVMConfig() *EVMRPCConfig {
+	return &EVMRPCConfig{
+		Enable:     true,
+		API:        "",
```
Comment (fedekunze) [inline cmd/ethermintd/config/config.go:78]: the default should include the `"eth"` namespace only

File: cmd/ethermintd/config/config.go [hunk 8]
```diff
@@ -95,6 +98,7 @@ func GetConfig(v *viper.Viper) Config {
 		Config: cfg,
 		EVMRPC: EVMRPCConfig{
 			Enable:     v.GetBool("evm-rpc.enable"),
+			API:        v.GetString("evm-rpc.api"),
```
Comment (fedekunze) [inline cmd/ethermintd/config/config.go:101]:
```suggestion
			API:        v.GetStringSlice("evm-rpc.api"),
```

File: docs/basics/json_rpc.md [hunk 9]
```diff
@@ -11,6 +11,16 @@ Check the JSON-RPC methods and namespaces supported on Ethermint. {synopsis}
...
+
+By default, only `eth` namespace is enabled. In order to enable other namespaces use flag `--evm-rpc.api`.
+
+```ethermintd start --evm-rpc.api txpool,personal,net,debug,web3```
```
Comment (fedekunze) [inline docs/basics/json_rpc.md:22]:
```suggestion
```ethermintd start --evm-rpc.api eth,txpool,personal,net,debug,web3```
```

Comment (fedekunze) [reply docs/basics/json_rpc.md:22]: I know it's enabled by default but I think it should be named in the example

File: ethereum/rpc/apis.go [hunk 10]
```diff
@@ -33,53 +32,80 @@ const (
...
+	apis = append(apis,
```
Comment (fedekunze) [inline ethereum/rpc/apis.go:43]: move to `switch` below, the default will be defined on the `DefaultConfig`

File: ethereum/rpc/apis.go [hunk 11]
```diff
@@ -33,53 +32,80 @@ const (
...
-			Namespace: DebugNamespace,
-			Version:   apiVersion,
-			Service:   debug.NewInternalAPI(ctx),
+			Service:   filters.NewPublicAPI(ctx.Logger, tmWSClient, evmBackend),
 			Public:    true,
 		},
+	)
+
+	for _, api := range selectedAPIs {
+		switch api {
```
Comment (fedekunze) [inline ethereum/rpc/apis.go]: add a default case that logs an error if the RPC namespace is invalid

File: ethereum/rpc/apis.go (hunk 2 above)
Comment (fedekunze) [reply ethereum/rpc/apis.go]: the other problem I see is adding a namespace multiple times (eg `evm-rpc.apis=web3,web3,txpool,txpool`). We should either validate the slice or remove any duplicates

File: docs/basics/json_rpc.md (hunk 1 above)
Comment (crypto-facs) [reply docs/basics/json_rpc.md]: Fixed!

File: init.sh [hunk 12]
```diff
@@ -84,4 +84,4 @@ if [[ $1 == "pending" ]]; then
 fi
 
 # Start the node (remove the --pruning=nothing flag if historical queries are not needed)
-ethermintd start --pruning=nothing $TRACE --log_level $LOGLEVEL --minimum-gas-prices=0.0001aphoton
+ethermintd start --pruning=nothing $TRACE --log_level $LOGLEVEL --minimum-gas-prices=0.0001aphoton --evm-rpc.api txpool,personal,net,debug,web3
```
Comment (fedekunze) [inline init.sh:87]:
```suggestion
ethermintd start --pruning=nothing $TRACE --log_level $LOGLEVEL --minimum-gas-prices=0.0001aphoton --evm-rpc.api eth,txpool,personal,net,debug,web3
```

File: ethereum/rpc/apis.go (hunk 2 above)
Comment (crypto-facs) [reply ethereum/rpc/apis.go]: And good catch 🙏 ! Fixed

File: server/start.go [hunk 13]
```diff
@@ -179,6 +180,7 @@ which accepts a path for the resulting pprof file.
 	cmd.Flags().String(flagGRPCWebAddress, serverconfig.DefaultGRPCWebAddress, "The gRPC-Web server address to listen on")
 
 	cmd.Flags().Bool(flagEVMRPCEnable, true, "Define if the gRPC server should be enabled")
+	cmd.Flags().String(flagEVMRPCAPI, "", "Defined which gRPC namespace should be enabled")
```
Comment (fedekunze) [inline server/start.go:183]: string slice, default value should be "eth"

Comment (fedekunze) [reply server/start.go:183]: you can also define a variable for the default values

File: tests/solidity/init-test-node.sh [hunk 14]
```diff
@@ -43,4 +43,4 @@ ethermintd collect-gentxs
 ethermintd validate-genesis
 
 # Start the node (remove the --pruning=nothing flag if historical queries are not needed)
-ethermintd start --pruning=nothing --rpc.unsafe --keyring-backend test --trace --log_level info
+ethermintd start --pruning=nothing --rpc.unsafe --keyring-backend test --trace --log_level info --evm-rpc.api txpool,personal,net,debug,web3
```
Comment (fedekunze) [inline tests/solidity/init-test-node.sh:46]:
```suggestion
ethermintd start --pruning=nothing --rpc.unsafe --keyring-backend test --trace --log_level info --evm-rpc.api eth,txpool,personal,net,debug,web3
```

**Review by fedekunze — CHANGES_REQUESTED** [review]
Looks good! minor comments

File: cmd/ethermintd/config/config.go (hunk 6 above)
Comment (crypto-facs) [reply cmd/ethermintd/config/config.go:67]: Hey Fede! This is a string, we then split it to get a slice out of it at [line](https://github.com/tharsis/ethermint/pull/349/files#diff-4296685a7348b4ed68d276fc1a5f35779fb132578be1760454aeeb3ceb4f2c89R28)

Comment (crypto-facs) [reply cmd/ethermintd/config/config.go:67]: Oh nevermind, just saw there is a function that returns this as slice by default! Will commit the suggestions

File: ethereum/rpc/apis.go [hunk 15]
```diff
@@ -33,53 +32,73 @@ const (
...
-			Service:   txpool.NewPublicAPI(ctx.Logger),
-			Public:    true,
-		},
-		{
-			Namespace: DebugNamespace,
-			Version:   apiVersion,
-			Service:   debug.NewInternalAPI(ctx),
-			Public:    true,
-		},
+	for index := range selectedAPIs {
```
Comment (fedekunze) [inline ethereum/rpc/apis.go:42]:
```suggestion
	for _, namespace := range selectedAPIs {
```

File: ethereum/rpc/apis.go [hunk 16]
```diff
@@ -33,53 +32,73 @@ const (
...
+		switch selectedAPIs[index] {
```
Comment (fedekunze) [inline ethereum/rpc/apis.go:43]:
```suggestion
		switch namespace {
```

File: ethereum/rpc/apis.go [hunk 17]
```diff
@@ -33,53 +32,73 @@ const (
...
+			apis = append(apis,
+				rpc.API{
+					Namespace: DebugNamespace,
+					Version:   apiVersion,
+					Service:   debug.NewInternalAPI(ctx),
+					Public:    true,
+				},
+			)
+		default:
+			ctx.Logger.Error("Invalid namespace value", "namespace", selectedAPIs[index])
```
Comment (fedekunze) [inline ethereum/rpc/apis.go:99]:
```suggestion
			ctx.Logger.Error("invalid namespace value", "namespace", namespace)
```

File: testutil/network/util.go [hunk 18]
```diff
@@ -121,7 +121,11 @@ func startInProcess(cfg Config, val *Validator) error {
 
 		val.jsonRPC = jsonrpc.NewServer()
 
-		apis := rpc.GetRPCAPIs(val.Ctx, val.ClientCtx, tmWsClient)
+		rpcAPI := val.AppConfig.EVMRPC.API
+		rpcAPI = strings.ReplaceAll(rpcAPI, " ", "")
+		rpcAPIArr := strings.Split(rpcAPI, ",")
```
Comment (fedekunze) [inline testutil/network/util.go]: not necessary anymore

File: server/evmrpc.go [hunk 19]
```diff
@@ -21,7 +22,11 @@ func StartEVMRPC(ctx *server.Context, clientCtx client.Context, tmRPCAddr string
 	tmWsClient := ConnectTmWS(tmRPCAddr, tmEndpoint)
 
 	rpcServer := ethrpc.NewServer()
-	apis := rpc.GetRPCAPIs(ctx, clientCtx, tmWsClient)
+
+	rpcAPI := config.EVMRPC.API
+	rpcAPI = strings.ReplaceAll(rpcAPI, " ", "")
+	rpcAPIArr := strings.Split(rpcAPI, ",")
```
Comment (fedekunze) [inline server/evmrpc.go]: not necessary anymore

File: CHANGELOG.md [hunk 20]
```diff
@@ -71,6 +71,8 @@ Ref: https://keepachangelog.com/en/1.0.0/
...
+* (rpc) [#349](https://github.com/tharsis/ethermint/pull/349) Implement configurable JSON-RPC APIs to manage enabled namespaces.
```
Comment (fedekunze) [inline CHANGELOG.md:75]: the line above is duplicated

**Review by fedekunze — APPROVED** [review]
LGTM. The only thing left is preventing adding the APIs namespaces more than once (convert slice to set)

File: docs/basics/json_rpc.md [hunk 21]
```diff
@@ -11,6 +11,16 @@ Check the JSON-RPC methods and namespaces supported on Ethermint. {synopsis}
 - [Ethereum JSON-RPC](https://eth.wiki/json-rpc/API) {prereq}
 - [Geth JSON-RPC APIs](https://geth.ethereum.org/docs/rpc/server) {prereq}
 
+## JSON-RPC Server
+
+To enable RPC server use the following flag (set to true by default).
+
+```ethermintd start --evm-rpc.enable```
```
Comment (fedekunze) [inline docs/basics/json_rpc.md:18]: this should be:
```bash
ethermintd start --evm-rpc.enable
```

File: docs/basics/json_rpc.md [hunk 22]
```diff
@@ -11,6 +11,16 @@ Check the JSON-RPC methods and namespaces supported on Ethermint. {synopsis}
...
+
+By default, only `eth` namespace is enabled. In order to enable other namespaces use flag `--evm-rpc.api`.
+
+```ethermintd start --evm-rpc.api eth,txpool,personal,net,debug,web3```
```
Comment (fedekunze) [inline docs/basics/json_rpc.md:22]: ditto break line

**Pull request merged**